### PR TITLE
Added fixNodeTempateID to update rancher2_node_template ID

### PR DESCRIPTION
**Pull request by rawmind0**
Related to https://github.com/rancher/rancher/pull/23718  
Fix issues #195  and https://github.com/rancher/rancher/issues/26160

### CODE REVIEW

**Comment by rawmind0** [comment]
Tested and working fine on rancher upgrade plan, from v2.2.10 -> v2.3.2 -> v2.3.3 -> v2.3.6 - > v2.4.0

TF file used to test node_template

```
resource "rancher2_cloud_credential" "foo" {
  name = "foo"
  description= "Terraform cloudCredential acceptance test"
  amazonec2_credential_config {
	access_key = "XXXXXXXXXXXXXXXXXXXX"
	secret_key = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  }
}

resource "rancher2_node_template" "foo" {
  name = "foo"
  description = "Terraform node driver amazonec2 acceptance test"
  cloud_credential_id = rancher2_cloud_credential.foo.id
  amazonec2_config {
	ami =  "ami-XXXXXXXXXXXXXXY"
	region = "XX-west-1"
	security_group = ["XXXXXXXX"]
	subnet_id = "subnet-XXXXXXXX"
	vpc_id = "vpc-XXXXXXXX"
	zone = "a"
  }
}
```

Test process

```
# Start standalone rancher v2.2.10 with shared volume and get rancher token
docker create -v /var/lib/rancher --name=rancher-storage rancher/rancher:v2.3.0
docker run -d --restart=unless-stopped -p 443:443 --volumes-from=rancher-storage --name=rancher-server rancher/rancher:v2.2.10

# Download terraform and built rancher2 provider repo, https://github.com/terraform-providers/terraform-provider-rancher2/pull/288 
git clone https://github.com/terraform-providers/terraform-provider-rancher2
cd terraform-provider-rancher2
git fetch origin pull/288/head:288
git checkout 288
make bin

# copy TF file to the current folder, config and execute terraform
export RANCHER_URL=https://127.0.0.1:443
export RANCHER_TOKEN_KEY=<rancher_token>
export RANCHER_INSECURE=true
terraform init
terraform apply

# This will create a cloud credential and a node template. Node template id should be on format "user-XXXXX:nt-XXXXX"
```

```
# Upgrade rancher to v2.3.2 (just previous version before ID update) and run terraform apply. TF Plan should be empty, no node template ID update
docker rm -f rancher-server
docker run -d --restart=unless-stopped -p 443:443 --volumes-from=rancher-storage --name=rancher-server rancher/rancher:v2.3.2
terraform apply
rancher2_cloud_credential.foo: Refreshing state... [id=cattle-global-data:foo]
rancher2_node_template.foo: Refreshing state... [id=user-XXXXX:nt-XXXXX]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

```
# Upgrade rancher to v2.3.3 (wait a couple of minutes until upgrade process is finished) and run terraform apply. TF should read and update node template ID on first run (from user-XXXXX:nt-XXXXX to  cattle-global-nt:nt-user-XXXXX-nt-XXXXX), but tf plan should be empty
docker rm -f rancher-server
docker run -d --restart=unless-stopped -p 443:443 --volumes-from=rancher-storage --name=rancher-server rancher/rancher:v2.3.3
terraform apply #fist run
rancher2_cloud_credential.foo: Refreshing state... [id=cattle-global-data:foo]
rancher2_node_template.foo: Refreshing state... [id=user-XXXXX:nt-XXXXX]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

terraform apply  #second run
rancher2_cloud_credential.foo: Refreshing state... [id=cattle-global-data:foo]
rancher2_node_template.foo: Refreshing state... [id=cattle-global-nt:nt-user-XXXXX-nt-XXXXX]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

```
# Upgrade rancher to v2.3.6 (wait a couple of minutes until upgrade process is finished) and run terraform apply. TF Plan should be empty, maintaining node template ID update
docker rm -f rancher-server
docker run -d --restart=unless-stopped -p 443:443 --volumes-from=rancher-storage --name=rancher-server rancher/rancher:v2.3.6
terraform apply
rancher2_cloud_credential.foo: Refreshing state... [id=cattle-global-data:foo]
rancher2_node_template.foo: Refreshing state... [id=cattle-global-nt:nt-user-XXXXX-nt-XXXXX]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

```
# Upgrade rancher to v2.4.0 (wait a couple of minutes until upgrade process is finished) and run terraform apply. TF Plan should be empty, maintaining node template ID update
docker rm -f rancher-server
docker run -d --restart=unless-stopped -p 443:443 --volumes-from=rancher-storage --name=rancher-server rancher/rancher:v2.4.0
terraform apply
rancher2_cloud_credential.foo: Refreshing state... [id=cattle-global-data:foo]
rancher2_node_template.foo: Refreshing state... [id=cattle-global-nt:nt-user-XXXXX-nt-XXXXX]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

File: rancher2/structure_node_template.go [hunk 1]
```diff
@@ -2,10 +2,27 @@ package rancher2
...
 
+func fixNodeTempateID(current, id string) string {
+	if len(current) == 0 || len(id) == 0 {
+		return ""
+	}
+	// Fix breaking API change https://github.com/rancher/rancher/pull/23718
+	limitVersion := "2.3.3"
+	if ok, _ := IsVersionGreaterThanOrEqual(current, limitVersion); ok {
+		prefix := "cattle-global-nt:nt-"
+		if !strings.HasPrefix(id, prefix) {
```
Comment (rmweir) [inline rancher2/structure_node_template.go]: I would maybe just say if it doesn't have prefix "cattle-global-nt:" to future proof. who knows if we'll ever make a nodetemplate without "nt-" prefix.

File: rancher2/config.go [hunk 2]
```diff
@@ -88,6 +88,28 @@ func (c *Config) getK8SVersion() (string, error) {
 	return c.K8SVer, nil
 }
 
+func (c *Config) IsRancherVersionLessThanl(ver string) (bool, error) {
```
Comment (rmweir) [inline rancher2/config.go]: is this function used? and the "l" at the end?

**Review by rmweir — CHANGES_REQUESTED** [review]

File: rancher2/config.go (hunk 2 above)
Comment (rawmind0) [reply rancher2/config.go]: Fixed function name

File: rancher2/structure_node_template.go (hunk 1 above)
Comment (rawmind0) [reply rancher2/structure_node_template.go]: What do you propose??

File: rancher2/structure_node_template.go [hunk 3]
```diff
@@ -2,10 +2,27 @@ package rancher2
...
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 )
 
+func fixNodeTempateID(current, id string) string {
+	if len(current) == 0 || len(id) == 0 {
+		return ""
+	}
+	// Fix breaking API change https://github.com/rancher/rancher/pull/23718
+	limitVersion := "2.3.3"
+	if ok, _ := IsVersionGreaterThanOrEqual(current, limitVersion); ok {
```
Comment (luthermonson) [inline rancher2/structure_node_template.go]: you added IsRancherVersionLessThan and IsRancherVersionGreaterThanOrEqual and then didn't use them and used your util function IsVersionGreaterThanOrEqual which you use in the two rancher version checkers too.. do one or the other maybe? you woulndt need to pass in `current` and could do everything inside this function. if you dont want to do that then maybe remove the two unused functions unless you have a purpose for them in the future.

Comment (rawmind0) [reply rancher2/structure_node_template.go]: You are right. Updated PR.

File: rancher2/resource_rancher2_node_template.go [hunk 4]
```diff
@@ -86,30 +86,32 @@ func resourceRancher2NodeTemplateCreate(d *schema.ResourceData, meta interface{}
 }
 
 func resourceRancher2NodeTemplateRead(d *schema.ResourceData, meta interface{}) error {
-	log.Printf("[INFO] Refreshing Node Template ID %s", d.Id())
+	id := d.Id()
```
Comment (luthermonson) [inline rancher2/resource_rancher2_node_template.go]: nit: you call d.SetId() later, not sure you need this local var. maybe just use the d.Id() everywhere you have this.

Comment (rawmind0) [reply rancher2/resource_rancher2_node_template.go]: ok, updated.

**Review by luthermonson — APPROVED** [review]

**Review by cloudnautique — APPROVED** [review]

**Pull request merged**